core/src/main/scala/kafka/log/Log.scala

```diff
@@ -541,7 +557,7 @@ class Log(@volatile var dir: File,
         info(s"Recovering unflushed segment ${segment.baseOffset}")
         val truncatedBytes =
           try {
-            recoverSegment(segment, Some(_leaderEpochCache))
+            recoverSegment(segment, if (supportsLeaderEpoch) Some(_leaderEpochCache) else None)
```

Comment:
@hachikuji 
I had to add this because we would still rebuild the cache in the case of an unclean shutdown and restart with `log.message.format=0.10`
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/LogSegment.scala#L378

Note that I removed the last test case in `testOldMessageFormatDeletesEpochCacheIfUnsupported()`. If we re-initialize the log with the new message format and unclean shutdown, it **will** rebuild it from the batches. This further exemplifies that we don't properly support message downgrades I think.

Can you think of any implications here? I believe we might be fine

Reply:
Yeah, it seems ok. We can open a separate JIRA for the problem you raised since it affects trunk as well.

Reply:
Yep. Opened https://issues.apache.org/jira/browse/KAFKA-7984